source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,360 @@
+  skipAcls(data, offset);
+  int32_t version;
+  BufferHelper::peekInt32(data, offset, version);
+
+  callbacks_.onSetAclRequest(path, version);
+}
+
+void DecoderImpl::parseSyncRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 8);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+
+  callbacks_.onSyncRequest(path);
+}
+
+void DecoderImpl::parseCheckRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 8);
+
+  std::string path;
+  BufferHelper::peekString(data, offset, path);
+  int32_t version;
+  BufferHelper::peekInt32(data, offset, version);
+
+  callbacks_.onCheckRequest(path, version);
+}
+
+void DecoderImpl::parseMultiRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  // Treat empty transactions as a decoding error, there should be at least 1 header.
+  CHECK_LENGTH(len, 17);
+
+  while (true) {
+    int32_t type;
+    BufferHelper::peekInt32(data, offset, type);
+    bool done{};
+    BufferHelper::peekBool(data, offset, done);
+    int32_t error;
+    BufferHelper::peekInt32(data, offset, error);
+
+    if (done) {
+      break;
+    }
+
+    switch (type) {
+    case enumToInt(OpCodes::CREATE):
+      parseCreateRequest(data, offset, len, false);
+      break;
+    case enumToInt(OpCodes::SETDATA):
+      parseSetRequest(data, offset, len);
+      break;
+    case enumToInt(OpCodes::CHECK):
+      parseCheckRequest(data, offset, len);
+      break;
+    default:
+      // Should not happen.
+      throw new EnvoyException("Unknown type within a transaction");
+    }
+  }
+
+  callbacks_.onMultiRequest();
+}
+
+void DecoderImpl::parseReconfigRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 28);
+
+  // Skip joining.
+  int32_t datalen;
+  BufferHelper::peekInt32(data, offset, datalen);
+  offset += datalen;
+  // Skip leaving.
+  BufferHelper::peekInt32(data, offset, datalen);
+  offset += datalen;
+  // Skip new members.
+  BufferHelper::peekInt32(data, offset, datalen);
+  offset += datalen;
+  // Read config id.
+  int64_t config_id;
+  BufferHelper::peekInt64(data, offset, config_id);
+
+  callbacks_.onReconfigRequest();
+}
+
+void DecoderImpl::parseSetWatchesRequest(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  CHECK_LENGTH(len, 12);
+
+  // Data watches.
+  skipStrings(data, offset);
+  // Exist watches.
+  skipStrings(data, offset);
+  // Child watches.
+  skipStrings(data, offset);
+
+  callbacks_.onSetWatchesRequest();
+}
+
+void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) const {
+  int32_t count;
+  BufferHelper::peekInt32(data, offset, count);
+
+  for (int i = 0; i < count; ++i) {
+    int32_t len;
+    BufferHelper::peekInt32(data, offset, len);
+    offset += len;
+  }
+}
+
+void DecoderImpl::onData(Buffer::Instance& data) {
+  uint64_t offset = 0;
+  while (offset < data.length()) {
+    try {
+      decode(data, offset);
```

Comment:
Any chance that offset will still be 0 after this throws? Maybe put the try catch around the while loop to avoid the potential infinite loop

Reply:
That's probably safer, to give up as soon as something goes wrong.